### PR TITLE
Preventing Open Redirect Attacks

### LINKED ISSUE #92: Preventing Open Redirect Attacks
ASP.NET Core -> Security -> Preventing Open Redirect Attacks

- What is an Open Redirect Attack?
  - Describe the problem
  - Show an example attack
- Protecting Against Open Redirect Attacks
  - Using Url.IsLocalUrl to protect redirection done by the server

Current default templates use this helper in `AccountController`:
```
        private IActionResult RedirectToLocal(string returnUrl)
        {
            if (Url.IsLocalUrl(returnUrl))
            {
                return Redirect(returnUrl);
            }
            else
            {
                return RedirectToAction(nameof(HomeController.Index), "Home");
            }
        }
```

[Edit Tom Dykstra] [commit with deleted article stub](https://github.com/aspnet/Docs/commit/e4940fb39e40b4af7973e0709c004c6a6cc62f21)

### PR DESCRIPTION
**Pull request by ardalis**
[Internal Review URL](https://review.docs.microsoft.com/en-us/aspnet/core/security/preventing-open-redirects?branch=pr-en-us-3619)
Fixes #92

### CODE REVIEW

File: aspnetcore/security/preventing-open-redirects.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+If you navigate to the link above and log in, you should be redirected to the site's [About page](http://nerddinner.com/Home/About). Imagine an attack designed to gain access to a user's credentials. The attacker would create a page that matched the look of the target app, and would most likely use a similar domain name as well (e.g. nerddiner.com, with a single 'n'). Then the attacker would get an unsuspecting user to navigate to a page requesting them to change their password, using a URL like this: ``http://nerddinner.com/Account/LogOn?returnUrl=http://nerddiner.com/Account/LogOn``. After successfully logging in, the user will see the page reload and they'll be prompted to log in once more. Most likely, they'll think their first attempt had a mistake, and they will try again. The attacker will now receive the user's credentials, and can redirect the user back to the original, legitimate site so the user is unaware the attack took place.
+
+![Open Redirection Attack Process](preventing-open-redirects/_static/open-redirection-attack-process.png)
```
Comment (Rick-Anderson) [inline aspnetcore/security/preventing-open-redirects.md]: I understand the preceding paragraphs but I don't understand the image. Delete unless you can convince  barry it's OK.

**Comment by ardalis** [comment]
Hopefully good to go; awaiting shipit from @HaoK

**Comment by HaoK** [comment]
:shipit:

File: aspnetcore/security/preventing-open-redirects.md (hunk 1 above)
Comment (ardalis) [inline aspnetcore/security/preventing-open-redirects.md]: @blowdart does the image make sense to you? About the only thing I can think that might be unclear is that the 301 redirect is the reason for the user being taken to the malicious site (at the bottom). I could add an arrow indicating that, perhaps.

Comment (blowdart) [inline aspnetcore/security/preventing-open-redirects.md]: What might be better is "Imagine you have a page with an open redirect on your site, /Home/Redirect/. An attacker could create, for example, a link in an email, http://nerddinner.com/Home/Redirect?url=http://phishingsite.com/Home/Login. A typical user will look at the URL and see it begins with your site name, trust it, and click it. The open redirect would then kick in, sending them to the phishing site which looks identical to yours, where they login to what they think is your site...

Comment (ardalis) [inline aspnetcore/security/preventing-open-redirects.md]: I think you're referring to the paragraph, not the image (we can update the paragraph as well, of course). The image is this one: https://github.com/aspnet/Docs/blob/d893c04df6062202b3bc4471d4043a3d6dc0666c/aspnetcore/security/preventing-open-redirects/_static/open-redirection-attack-process.png

**Comment by ardalis** [comment]
@Rick-Anderson I updated the figure - is it more clear to you now?

**Comment by Rick-Anderson** [comment]
Can we work in [@]blowdart suggestion? I think he wants that for an example of an attack.
>What might be better is "Imagine you have a page with an open redirect on your site, /Home/Redirect/. An attacker could create, for example, a link in an email, http://nerddinner.com/Home/Redirect?url=http://phishingsite.com/Home/Login. A typical user will look at the URL and see it begins with your site name, trust it, and click it. The open redirect would then kick in, sending them to the phishing site which looks identical to yours, where they login to what they think is your site...

No, the flow doesn't make any sense to me. Can you outline the flow?

**Comment by ardalis** [comment]
The image outlines the flow.
1. Click a link to nerddinner.com/account/login?redirectUrl=http://nerddiner.com/account/login
2. Log in successfully
3. Redirected to http://nerddiner.com/account/login (malicious site that looks like initial site)
4. Log in again (giving malicious site your credentials); redirected back to real site

This is just one example. The basic idea is that a site that accepts any redirect URL as a querystring can easily be made to redirect a user to a site masquerading as the site they trust, and they won't suspect it since the initial URL was, in fact, their trusted site's.

**Comment by Rick-Anderson** [comment]
Can you add that list?
1. Click a link to `nerddinner.com/account/login?redirectUrl=http://nerddiner.com/account/login`  (nerddi**NN**er.com is the expected site, nerddi**N**er.com is the malicious site.)
2. Log in successfully
3. Redirected to `http://nerddiner.com/account/login` (malicious site that looks like real site).
4. Log in again (giving malicious site your credentials); redirected back to real site.

**Comment by Rick-Anderson** [comment]
Be sure to work in Barry's comment.

**Comment by ardalis** [comment]
how's this look now?

**Pull request merged**